### ISSUE TITLE
Create/update subscriptions in bulk

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus#4632]
Current Subscription API caters for individual subscriptions (create/delete).

During startup, this information could be passed in as a collection to allow transport to optimize the startup time by leveraging smaller number of management operations.

Original (user) request: https://github.com/Particular/NServiceBus.AzureServiceBus/issues/522

**Comment by timbussmann** [comment]
so essentially a `messageSession.Subscribe(eventType1, eventType2, eventType3, ...)` is what would improve your situation?

But it seems that it would be a cleaner approach to follow up on the "batchable sends" ideas for the message session in general, as this the above api would allow batching for subscriptions but not for sends.

(would a `messageSession.Send(message1, message2, message3)` be the simplest way to implement the batching instead of considering session lifetimes?) @andreasohlund @danielmarbach

**Comment by SeanFeldman** [comment]
Not quite.
What it would mean is to have an addition to `IManageSubscriptions` that has a `Subscribe(List<Type> events,  ContextBag context)` that is invoked when endpoint is bootstrapped and events are scanned. Doing it at the time of when messages are send (i.e. `messageSession.Send(message1, message2, message3)`) is too late.

**Comment by andreasohlund** [comment]
We have talked about passing the list of subscribers to the transport at
.Init time. That would be better imo since allows the transport to abort
startup should something go wrong when the subscriptions are setup?

**Comment by timbussmann** [comment]
so the core currently only subscribes via the `AutoSubscribe` startup task which is done when the endpoint is essentially started but before it starts consuming messages. Passing subscribers earlier would probably be quite a change and only works with native pub/sub transports and not with message driven pub/sub (as it requires a started transport by it's nature) so that sounds like a much larger change compared to my proposal?

@SeanFeldman how would you call the `IManageSubscriptions` new `BulkSubscribe` method without an matching API on the message session?
The mention of send was not related to the bulk subscription topic but to bulk operations in general.

**Comment by SeanFeldman** [comment]
Due to the nature of those events (auto-subscribed upon endpoint startup), possible candidates:
- `SubscribeAutomatically`
- `SubscribeOnStartup`
- `SubscribeToKnownEvents`

**Comment by andreasohlund** [comment]
I was talking about the endgame, sorry should have been more clear. In a
world where transports fully manage the pubsub part I don't see how the
transport not being started would be an issue.

Not against considering the bulk subscribe idea in the interim

**Comment by timbussmann** [comment]
@andreasohlund yeah, long term I agree with you, but that seems something for v8?

@SeanFeldman sorry for being unclear, I meant how would you call this method from code without a matching API in the message session? :)

**Comment by SeanFeldman** [comment]
@timbussmann there's some sort of confusion. What I'm referring to is not related to the message session, but is more aligned with a long term (what @andreasohlund is talking about). This would not be a user triggered operation.

**Comment by bording** [comment]
@SeanFeldman Take a look at how AutoSubscribe is implemented: https://github.com/Particular/NServiceBus/blob/127b10eb48efcc90eb4d5de7d9645535445d8aeb/src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs#L71

It only has access to an `IMessageSession`, so it just sends multiple `Subscribe` messages. The outgoing subscription pipeline is invoked for each subscription. The terminator of that pipeline is either the [MessageDrivenSubscribeTerminator ](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs)or the [NativeSubscribeTerminator](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Routing/NativeSubscribeTerminator.cs).

Because of how this works, there isn't currently a way to batch those subscriptions without adding a way to send a batch of them at once through `IMessageSession`.

**Comment by timbussmann** [comment]
exactly what @bording said. And if we talk about future major API changes I think there are two different things mentioned here:
* Generally rework the subscription mechanism towards an option which works nicer with native mechanisms (e.g. somehow on transport initialization as mentioned by @andreasohlund )
* Generally rework the message session API to allow batched operations. This is a topic already hugely discussed in Bangkok. Of course, this would/should more benefit sends/publishes in the first place, but it could also be used for subscriptions in case option 1 isn't already done by then. I actually wonder whether the `Send(params messages[])` approach wouldn't be a first, simple step towards batching support (yes, is extends the scope of this issue).

**Comment by SeanFeldman** [comment]
First, thank you @bording for visualizing the discussion. It helps.

Second, for the ASB there's a workaround for now (more management operations on the publisher side of things, but that should suffice), until Core API can be updated in the future.

**Comment by danielmarbach** [comment]
@timbussmann I'm not yet convinced that a Send with params is a good simple step towards batching. In the UoW proposal I've outlined a good and solid API with examples for various frameworks which would be a step towards batch and easily allows to hook those batched sessions on things like MVC controllers. Batching itself as I've proved with a few spikes is not that complex to achieve even if core only changes. So I'm not sure why simple needs to be an argument to go with the Send and params based overload. Happy to discuss this further but I think we should not go into this direction just yet.

**Comment by timbussmann** [comment]
discussed this on the @Particular/nservicebus-maintainers sync and decided to add it to the future milestone as it's a valid request but would require a major version as some kind of breaking changes would be involved. It seems there are multiple potential approaches which would include this feature, e.g.
* Change subscription mechanisms to be more transport specific and not use the pipeline in the first place
* Involve bulk operations in general (e.g. bulk sends)
